Add streaming version of XxHash64

electrum opened the pull request.

martint commented on src/main/java/io/airlift/slice/XxHash64.java:
```diff
@@ -69,8 +69,8 @@ public static long hash(long seed, Slice data, int offset, int length)
 
         // round to the closest 32 byte boundary
-        // this is the point up to which {@see #updateBody} processed
-        int index = length & 0xFFFFFFE0;
```
Don't do this. This is a mask, and `-32` makes it hard to tell what it is.

electrum replied:
I think is the standard way to write it and is more clear (the JDK does it this way), but I can leave it.

martint commented on src/main/java/io/airlift/slice/XxHash64.java:
```diff
@@ -56,7 +56,7 @@ public static long hash(long seed, Slice data, int offset, int length)
         checkPositionIndexes(0, offset + length, data.length());
 
         Object base = data.getBase();
-        final long address = data.getAddress() + offset;
```
This was intentional, to make sure the address variable is not inadvertently overwritten as the code walks loops through the data.

martint commented:
Have you run the benchmarks to make sure your changes didn't affect performance for the non-streaming version? The structure of the code was very carefully tuned for play nice with how hotspot inlines/optimizes it.

martint commented:
I'm not sure I like mixing the static vs non-static methods in the same class. I'd probably introduce a separate class for the streaming version

martint commented:
looks good

electrum commented:
I found that computing the body from `v1`-`v4` in a method caused a substantial performance penalty, so I removed that. Extracting a method for `updateTail` was the same performance. I moved that into its own commit to make it easy to benchmark in isolation.
